File: apps/frontend/src/adapters/shared/private-investments.ts
import { invoke } from "./platform";

export type CapitalCallStatus = "expected" | "due" | "paid" | "cancelled";

export interface PrivateInvestment {
  assetId: string;
  manager: string;
  strategy: string;
  vintageYear?: number | null;
  commitmentAmount: string;
  commitmentCurrency: string;
  inceptionDate?: string | null;
  notes?: string | null;
}

export interface PrivateInvestmentValuation {
  id: string;
  assetId: string;
  valuationDate: string;
  nav: string;
  currency: string;
  sourceCitationId?: string | null;
}

export interface CapitalCall {
  id: string;
  assetId: string;
  noticeDate: string;
  dueDate: string;
  amount: string;
  currency: string;
  status: CapitalCallStatus;
  sourceCitationId?: string | null;
  notes?: string | null;
}

export interface PrivateDistribution {
  id: string;
  assetId: string;
  distributionDate: string;
  amount: string;
  currency: string;
  recallable: boolean;
  sourceCitationId?: string | null;
  notes?: string | null;
}

export interface PrivateInvestmentSummary {
  investment: PrivateInvestment;
  commitment: string;
  paidInCapital: string;
  unfundedCommitment: string;
  totalDistributions: string;
  recallableDistributions: string;
  currentNav: string;
  dpi?: string | null;
  rvpi?: string | null;
  tvpi?: string | null;
  moic?: string | null;
  warnings: string[];
}

export interface UpsertPrivateInvestmentRequest {
  assetId: string;
  manager: string;
  strategy: string;
  vintageYear?: number | null;
  commitmentAmount: string;
  commitmentCurrency: string;
  inceptionDate?: string | null;
  notes?: string | null;
}

export interface CreatePrivateInvestmentValuationRequest {
  assetId: string;
  valuationDate: string;
  nav: string;
  currency: string;
  sourceCitationId?: string | null;
}

export interface CreateCapitalCallRequest {
  assetId: string;
  noticeDate: string;
  dueDate: string;
  amount: string;
  currency: string;
  status: CapitalCallStatus;
  sourceCitationId?: string | null;
  notes?: string | null;
}

export interface UpdateCapitalCallStatusRequest {
  id: string;
  status: CapitalCallStatus;
}

export interface CreatePrivateDistributionRequest {
  assetId: string;
  distributionDate: string;
  amount: string;
  currency: string;
  recallable: boolean;
  sourceCitationId?: string | null;
  notes?: string | null;
}

export function upsertPrivateInvestment(
  request: UpsertPrivateInvestmentRequest,
): Promise<PrivateInvestment> {
  return invoke<PrivateInvestment>("upsert_private_investment", { request });
}

export function getPrivateInvestment(assetId: string): Promise<PrivateInvestment | null> {
  return invoke<PrivateInvestment | null>("get_private_investment", { assetId });
}

export function deletePrivateInvestment(assetId: string): Promise<void> {
  return invoke<void>("delete_private_investment", { assetId });
}

export function addPrivateInvestmentValuation(
  request: CreatePrivateInvestmentValuationRequest,
): Promise<PrivateInvestmentValuation> {
  return invoke<PrivateInvestmentValuation>("add_private_investment_valuation", { request });
}

export function addCapitalCall(request: CreateCapitalCallRequest): Promise<CapitalCall> {
  return invoke<CapitalCall>("add_capital_call", { request });
}

export function updateCapitalCallStatus(
  request: UpdateCapitalCallStatusRequest,
): Promise<CapitalCall> {
  return invoke<CapitalCall>("update_capital_call_status", { request });
}

export function addPrivateDistribution(
  request: CreatePrivateDistributionRequest,
): Promise<PrivateDistribution> {
  return invoke<PrivateDistribution>("add_private_distribution", { request });
}

export function getPrivateInvestmentSummary(
  assetId: string,
): Promise<PrivateInvestmentSummary | null> {
  return invoke<PrivateInvestmentSummary | null>("get_private_investment_summary", { assetId });
}

File: apps/tauri/src/commands/private_investments.rs
use std::sync::Arc;

use log::error;
use mizan_core::private_investments::{
    CapitalCall, CreateCapitalCallRequest, CreatePrivateDistributionRequest,
    CreatePrivateInvestmentValuationRequest, PrivateDistribution, PrivateInvestment,
    PrivateInvestmentRepositoryTrait, PrivateInvestmentSummary, PrivateInvestmentValuation,
    UpdateCapitalCallStatusRequest, UpsertPrivateInvestmentRequest,
};
use tauri::State;

use crate::context::ServiceContext;

#[tauri::command]
pub async fn upsert_private_investment(
    request: UpsertPrivateInvestmentRequest,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<PrivateInvestment, String> {
    state
        .private_investment_repository()
        .upsert_investment(request)
        .await
        .map_err(command_error("upsert_private_investment"))
}

#[tauri::command]
pub async fn get_private_investment(
    asset_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Option<PrivateInvestment>, String> {
    state
        .private_investment_repository()
        .get_investment(&asset_id)
        .await
        .map_err(command_error("get_private_investment"))
}

#[tauri::command]
pub async fn delete_private_investment(
    asset_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<(), String> {
    state
        .private_investment_repository()
        .delete_investment(&asset_id)
        .await
        .map_err(command_error("delete_private_investment"))
}

#[tauri::command]
pub async fn add_private_investment_valuation(
    request: CreatePrivateInvestmentValuationRequest,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<PrivateInvestmentValuation, String> {
    state
        .private_investment_repository()
        .add_valuation(request)
        .await
        .map_err(command_error("add_private_investment_valuation"))
}

#[tauri::command]
pub async fn add_capital_call(
    request: CreateCapitalCallRequest,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<CapitalCall, String> {
    state
        .private_investment_repository()
        .add_capital_call(request)
        .await
        .map_err(command_error("add_capital_call"))
}

#[tauri::command]
pub async fn update_capital_call_status(
    request: UpdateCapitalCallStatusRequest,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<CapitalCall, String> {
    state
        .private_investment_repository()
        .update_capital_call_status(request)
        .await
        .map_err(command_error("update_capital_call_status"))
}

#[tauri::command]
pub async fn add_private_distribution(
    request: CreatePrivateDistributionRequest,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<PrivateDistribution, String> {
    state
        .private_investment_repository()
        .add_distribution(request)
        .await
        .map_err(command_error("add_private_distribution"))
}

#[tauri::command]
pub async fn get_private_investment_summary(
    asset_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Option<PrivateInvestmentSummary>, String> {
    state
        .private_investment_repository()
        .get_summary(&asset_id)
        .await
        .map_err(command_error("get_private_investment_summary"))
}

fn command_error(command: &'static str) -> impl FnOnce(mizan_core::Error) -> String {
    move |err| {
        error!("{command} failed: {err}");
        format!("{command} failed: {err}")
    }
}
